pinot-core/src/test/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManagerTest.java

```diff
@@ -609,6 +639,43 @@ public void testFileRemovedDuringOnlineTransition()
     segmentDataManager._state.set(segmentDataManager, LLRealtimeSegmentDataManager.State.HOLDING);
     segmentDataManager.goOnlineFromConsuming(metadata);
     Assert.assertFalse(new File(segTarFileName).exists());
+    segmentDataManager.destroy();
+  }
+
+  @Test
+  public void testOnlyOneSegmentHoldingTheSemaphoreForParticularPartition()
+      throws Exception {
+    long timeout = 2_000L;
+    FakeLLRealtimeSegmentDataManager firstSegmentDataManager = createFakeSegmentManager();
+    Assert.assertTrue(firstSegmentDataManager.getAcquireConsumerSemaphore().get());
+    Assert.assertEquals(firstSegmentDataManager.getPartitionConsumerSemaphore().availablePermits(), 0);
+
+    // Release semaphore after timeout.
+    Thread releaseSemaphore = new Thread(() -> {
+      try {
+        Thread.sleep(timeout);
+      } catch (InterruptedException e) {
+        throw new RuntimeException("InterruptedException when sleeping for " + timeout + "ms");
+      } finally {
+        firstSegmentDataManager.destroy();
+      }
+    });
+    releaseSemaphore.start();
+
+    // The second segment will be hanged until the semaphore gets released.
+    FakeLLRealtimeSegmentDataManager secondSegmentDataManager = createFakeSegmentManager();
+    Assert.assertEquals(firstSegmentDataManager.getPartitionConsumerSemaphore(),
+        secondSegmentDataManager.getPartitionConsumerSemaphore());
+    Assert.assertTrue(secondSegmentDataManager.getAcquireConsumerSemaphore().get());
```

Comment:
Perhaps I am missing something, but the second segment data manager should not have acquired the semaphore, so this should be asserted to false?

Reply:
The acquire method of semaphore is a blocking method. The second segmentManager will be blocked on the acquire method until the first segmentManager releases the permit. 
Thus, after the second segmentManager finishes its construction, it has already taken the permit. So it should be true here.